### PR TITLE
Fix some bugs in date time processing

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

## What have you changed? (mandatory)

This PR fixes some bugs preventing us to treat time error as warning **in some cases**.

Also changed to accept reference when formatting date time.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

No need.

### CODE REVIEW

File: src/coprocessor/codec/mysql/time/mod.rs [hunk 1]
```diff
@@ -549,151 +543,196 @@ impl Time {
...
                 }
             }
-            'Y' => Ok(format!("{:04}", self.time.year())),
+            'Y' => {
+                write!(output, "{:04}", self.time.year()).unwrap();
+            }
             'y' => {
                 let year_str = format!("{:04}", self.time.year());
-                Ok(year_str[2..].to_string())
+                write!(output, "{}", &year_str[2..]).unwrap();
```
Comment (kennytm) [inline src/coprocessor/codec/mysql/time/mod.rs]: push_str

(But why not write `year % 100`?)

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -26,11 +22,16 @@ impl ScalarFunc {
     ) -> Result<Option<Cow<'a, [u8]>>> {
         let t: Cow<'a, Time> = try_opt!(self.children[0].eval_time(ctx, row));
         if t.invalid_zero() {
-            return handle_incorrect_datetime_error(ctx, t).map(|_| None);
+            return ctx
+                .handle_invalid_time_error(Error::incorrect_datetime_value(&format!("{}", t)))
```
Comment (kennytm) [inline src/coprocessor/dag/expr/builtin_time.rs]: `&t.to_string()` ?

But given the repetition I'd create a function like 

```rust
fn ensure_valid(ctx: ..., t: &Time) -> Result<()> {
    if t.invalid_zero() { return ... }
    Ok(())
}

...

ensure_valid(ctx, &t)?;
```

Comment (breeswish) [reply src/coprocessor/dag/expr/builtin_time.rs]: We will refine errors later and the code shall be simplified then :)

File: src/coprocessor/codec/mysql/time/mod.rs [hunk 3]
```diff
@@ -549,151 +543,195 @@ impl Time {
...
+    fn write_date_format_segment(&self, b: char, output: &mut String) -> Result<()> {
         match b {
             'b' => {
                 let m = self.time.month();
                 if m == 0 || m > 12 {
-                    Err(box_err!("invalid time format"))
+                    return Err(box_err!("invalid time format"));
                 } else {
-                    Ok(MONTH_NAMES_ABBR[(m - 1) as usize].to_string())
+                    write!(output, "{}", MONTH_NAMES_ABBR[(m - 1) as usize]).unwrap();
```
Comment (kennytm) [inline src/coprocessor/codec/mysql/time/mod.rs]: This one is also `push_str` 😅

File: src/coprocessor/codec/mysql/time/mod.rs [hunk 4]
```diff
@@ -549,151 +543,195 @@ impl Time {
...
                 }
             }
             'M' => {
                 let m = self.time.month();
                 if m == 0 || m > 12 {
-                    Err(box_err!("invalid time format"))
+                    return Err(box_err!("invalid time format"));
                 } else {
-                    Ok(MONTH_NAMES[(m - 1) as usize].to_string())
+                    write!(output, "{}", MONTH_NAMES[(m - 1) as usize]).unwrap();
```
Comment (kennytm) [inline src/coprocessor/codec/mysql/time/mod.rs]: ditto

File: src/coprocessor/codec/mysql/time/mod.rs (hunk 3 above)
Comment (breeswish) [reply src/coprocessor/codec/mysql/time/mod.rs]: Sharp eye!

**Comment by breeswish** [comment]
@lonng @sticnarf PTAL, thanks

**Pull request merged**